client/asset/btc/btc.go

```diff
@@ -4347,6 +4347,104 @@ func (btc *baseWallet) createWitnessSig(tx *wire.MsgTx, idx int, pkScript []byte
 	return sig, privKey.PubKey().SerializeCompressed(), nil
 }
 
+// EstimateSendTxFee returns a tx fee estimate for sending or withdrawing the
+// provided amount using the provided feeRate.
+func (btc *baseWallet) EstimateSendTxFee(address string, sendAmount, feeRate uint64, subtract bool) (fee uint64, err error) {
+	if sendAmount == 0 {
+		return 0, fmt.Errorf("cannot check fee: send amount = 0")
+	}
+
+	_, err = btc.decodeAddr(address, btc.chainParams)
+	if err != nil {
+		return 0, fmt.Errorf("invalid address: %s", address)
+	}
+
+	feeRate = btc.feeRateWithFallback(feeRate)
+
+	minTxSize := uint64(dexbtc.MinimumTxOverhead + btc.outputSize())
+
+	// If subtract is true, select enough inputs for sendAmount. Fees will be taken
+	// from the sendAmount. If not, select enough inputs to cover minimum fees.
+	enough := func(inputsSize, sum uint64) bool {
+		if subtract {
+			return sum >= sendAmount
+		}
+		minFee := (minTxSize + inputsSize) * feeRate
+		return sum >= sendAmount+minFee
+	}
+
+	btc.fundingMtx.RLock()
+	utxos, _, _, err := btc.spendableUTXOs(0)
+	btc.fundingMtx.RUnlock()
+	if err != nil {
+		return 0, err
+	}
+	sum, inputsSize, _, _, _, _, err := fund(utxos, enough)
+	if err != nil {
+		return 0, err
+	}
+
+	txSize := minTxSize + uint64(inputsSize)
+	estFee := feeRate * txSize
+	remaining := sum - sendAmount
+
+	// Check if there will be a change output if there is enough remaining.
+	changeFee := uint64(btc.outputSize()) * feeRate
+	changeValue := remaining - estFee - changeFee
+	if subtract {
+		// fees are already included in sendAmount, anything else is change.
+		changeValue = remaining
+	}
+
+	var finalFee uint64
+	if isDust, err := btc.changeIsDust(changeValue, feeRate); err != nil {
+		return 0, err
+	} else if isDust {
+		// remaining cannot cover a non-dust change and the fee for the change.
+		finalFee = estFee + remaining
+	} else {
+		// additional fee will be paid for non-dust change
+		finalFee = estFee + changeFee
+	}
+
+	// Retrieve balance and check if wallet has enough to cover sendAmount and
+	// finalFee.
+	bal, err := btc.Balance()
+	if err != nil {
+		return 0, err
+	}
+
+	if !subtract && bal.Available < sendAmount+finalFee {
+		return 0, fmt.Errorf("insufficient balance to cover total spend, available %d < %d total to spend",
+			bal.Available, sendAmount+finalFee)
+	}
+	return finalFee, nil
+}
+
+// outputSize returns the size for a single output.
+func (btc *baseWallet) outputSize() int {
+	if btc.segwit {
+		return dexbtc.P2WPKHOutputSize
+	}
+	return dexbtc.P2PKHOutputSize
+}
+
+// changeIsDust checks if changeValue is dust value using the provided feeRate.
+func (btc *baseWallet) changeIsDust(changeValue uint64, feeRate uint64) (isDust bool, err error) {
+	// Grab a change addresss.
+	changeAddr, err := btc.node.changeAddress()
```

Comment:
We avoid generating a new address to check if (value, feeRate) would be a dust output.

It's unfortunate the current `IsDust` functions need an actual `wire.TxOut` instead of the info that actually matters (serialized size, segwit bool, output value).  We can either factor out the meat of [`IsDust`](https://github.com/ukane-philemon/dcrdex/blob/ca277d9a9953852df51682651487f068118a0652/dex/networks/btc/script.go#L382) to work with the values I just listed, or we can use a dummy `PkScript` of the correct format depending on `btc.segwit`